### ISSUE TITLE
Add ability to cancel uploads in processFile

### ISSUE THREAD

**Issue by raystorm** [aws-amplify/amplify-ui#5099]
### On which framework/platform would you like to see this feature implemented?

React

### Which UI component is this feature-request for?

Storage (Storage Manager)

### Please describe your feature-request in detail.

I am building an Enterprise Content Management System / Digital Asset Management System with Amplify.  
I am storing files in an S3 bucket. I have a page, that lets users fill out metadata related to a file, and upload the file. 
I am adding a check to see if a file with the same name already exists in the bucket or not.  If the file already exists, I want to cancel the upload and report an error message to the user, so they don't accidentally overwrite the existing file.  

I am running the following amplify versions:
```json
"dependencies": {
  "@aws-amplify/core": "^5.8.11",
  "@aws-amplify/ui-react": "^5.3.3",
  "@aws-amplify/ui-react-storage": "^2.3.3",
  "aws-amplify": "^5.3.17",
}
```

**Note:**  
I have not filed a bug report for the following issues I found while trying to programmatically cancel a running upload.
  1.  Rejecting the promise from `processFile()` bubbles all they way to the top of the app, like a JSX syntax error does.
  2. `clearFiles()`  Clears the `FileList` but does not fire any of the associated events for removing the files or canceling uploads.  `onRemoveUpload() && onCancelUpload()`

Let me know if I should file Bug Reports for those issues I found.
-------

**TLDR**:  I want to be able to cancel currently running file upload in `processFile()` if file checks fail.

### Please describe a solution you'd like.

In the ` async processFile()` Handler,   I would like to return reject the promise.
```
return Promise.Reject('File already exists');
```

------
**Alternate Solution**
Use a reference to cancel all Uploads similar to `clearFiles();`
```
ref.current.cancelUploads();
```

-----
**Alternate Solution 2**  
Disable Auto Upload, and programmatically start the upload in `processFile()`.
`processFile()` doesn't run until the upload is started, even if I could figure out how to call the start button.

-----
**Alternate Solution 3**  
Call the `onCancelUpload()` or `onRemoveUpload()` events for the file.  
They are not callable from the `processFile` object, and I couldn't figure out how to access them from the component.

### We love contributors! Is this something you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request.
- [ ] ⚠️ This feature might incur a breaking change.

**Comment by esauerbo** [comment]
Hey @raystorm thanks for opening this. We're adding this to our roadmap to consider for future feature work.

**Comment by raystorm** [comment]
I initially didn't check the `👋 I may be able to implement this feature request.` box, because I had other work in my own backlog ahead of this, and I wasn't sure I understood `@aws-amplify/ui-react-storage`'s implementation enough to attempt a fix.  
I have cleared my own backlog, to the point where I am ready to look into this.  
I'm still exploring the Logic of how to build and test `@aws-amplify/ui-react-storage` but since I have started exploring this, I figured I should come back and update this issue.

**Comment by lwang-79** [comment]
I encountered the same problem, so I have to use [DragZone](https://ui.docs.amplify.aws/react/components/dropzone#opening-the-file-picker) component instead of the StorageManager.

**Comment by varamsky** [comment]
I am facing a similar issue here. I need to prevent duplicate files from being accepted by the StorageManager. I want to handle it in the `processFile()` function but `Promise.reject()` throws an error. 

If I accept duplicate files then I need to give the user ability to delete the duplicate from the file list. But as the `"key"`(name of the files) is same for the original and duplicate they create an issue and lead to an inconsistent UI.

Can anyone please suggest a way to fix this?

P.S.:- I am aware that a new version has been released but I am currently stuck on 

```
"@aws-amplify/ui-react": "^5.3.0",
"@aws-amplify/ui-react-storage": "^2.3.1",
"aws-amplify": "^5.3.10"
```

**Comment by varamsky** [comment]
I also have these questions:-

1. Also, I see that the ref only exposes `clearFiles()` function. Is there a way in React to gain access to the DOM element or create a wrapper around `StorageManager` to expose and use more functions from the `ref` and thus have more access on the files list internally managed by `StorageManager`.

2. If none of the above is possible then is this already solved in a later version. For example in the new `FileUploader` component? Although, it doesn't seem that even the newer versions expose more methods from the `ref.
P.S. - not that I have the luxury to move to newer versions but if this is true then I can have a better argument with the team to move to newer version in the future.

**Comment by pranavosu** [comment]
Hi @varamsky is there a reason why you're stuck on an older version? this is still in our backlog, so `FileUploader` does not support rejected Promises yet. Any new change/fix would not be back-ported to older versions, so I'd recommend updating. 

That said, what additional methods would you need in the `ref` ? It seems like handling rejected Promises better in `processFiles` would satisfy the original use case.

**Comment by varamsky** [comment]
Hi @pranavosu 

Regarding the `ref` question - I was trying other hacks to get this done which why I mentioned it earlier. But if you can add the functionality to reject a file upload on the `processFile()` function then this should do the job. Please ignore ref in that case. (**Assumption** here is that when we reject a file upload, it will also make sure not to show the rejected file upload in the file list in the UI. This is probably a no-brainer but just mentioning it.)

We are stuck on an older version as it was just a hassle to move to newer version. If this is implemented in the new version `FileUploader` component then it would give us yet another reason to go ahead and switch to the new version as soon as possbile.

**Comment by pranavosu** [comment]
Got it. Thanks @varamsky, I'll let you know when I have an update.

**Comment by raystorm** [comment]
@varamsky, I tried for 6 months to get a PR accepted.  
Good Luck, I asked for this feature for the same reason BTW. I want to stop duplicate uploads with the same key.

**Comment by varamsky** [comment]
Thanks for the Good Luck wish @raystorm I really would need it I guess. I have seen the work that you have put it. It shouldn't be this difficult to finalize and merge a PR!
